src/libpmem2/mcsafe_ops_posix.c

```diff
@@ -64,35 +136,91 @@ pmem2_source_mcsafe_operation(struct pmem2_source *src, void *buf, size_t size,
 
 	/* fsdax read/write relies on pread/pwrite */
 	if (ftype == PMEM2_FTYPE_REG) {
-		ssize_t retsize = 0;
-		if (op_type == MCSAFE_OP_READ)
-			retsize = pread(fd, buf, size, (off_t)offset);
-		else if (op_type == MCSAFE_OP_WRITE)
-			retsize = pwrite(fd, buf, size, (off_t)offset);
-
+		ssize_t retsize = mcsafe_op_args[op_type].op(fd, buf, size,
+				(off_t)offset);
 		if (retsize == -1) {
 			if (errno == EIO) {
 				ERR("physical I/O error occured, " \
-					"possible bad block encountered");
+					"possible bad block");
 				return PMEM2_E_IO_FAIL;
 			}
 
-			if (op_type == MCSAFE_OP_READ)
-				ERR("!pread");
-			else if (op_type == MCSAFE_OP_WRITE)
-				ERR("!pwrite");
+			ERR("%s", mcsafe_op_args[op_type].op_err_str);
 			return PMEM2_E_ERRNO;
 		}
 
 		return 0;
 	}
 
-	/* TODO: devdax implementation of this function */
-	if (ftype == PMEM2_FTYPE_DEVDAX) {
-		ERR("operation doesn't supported devdax file type yet");
-		return PMEM2_E_NOSUPP;
+	/* devdax requires a mapping and sigbus handler (badblock error) */
+	ret = pmem2_config_new(&cfg);
```

Comment:
Is this code below only for devdax? Shouldn't it be in function dedicated for dev dax?